.github/workflows/ci.yaml

```diff
@@ -17,37 +20,107 @@ jobs:
         uses: styfle/cancel-workflow-action@0.11.0
         with:
           access_token: ${{ github.token }}
+
+  check_image:
+    name: Check if image exists in registry
+    runs-on: ubuntu-latest
+    outputs:
+      imageexists:  ${{ steps.check_image.outputs.imageexists }}
+      imagetag:  ${{ steps.image_hash.outputs.imagetag }}
+      imageexists_sles:  ${{ steps.check_image.outputs.imageexists_sles }}
+      imagetag_sles:  ${{ steps.image_hash.outputs.imagetag_sles }}
+
+    steps: 
+      - name: Checkout Code
+        uses: actions/checkout@v3
+
+      - name: Create Image Tag
+        id: image_hash
+        run: |
+          echo "imagetag=rocm/migraphx-private:hip-clang-${{hashFiles('**/hip-clang.docker', '**/*requirements.txt', '**/install_prereqs.sh', '**/rbuild.ini')}}" >> $GITHUB_OUTPUT
+          echo "imagetag_sles=rocm/migraphx-private:hip-clang-sles-${{hashFiles('**/tools/docker/sles.docker', '**/*requirements.txt', '**/install_prereqs.sh', '**/rbuild.ini')}}" >> $GITHUB_OUTPUT
+
+      - name: Check if image is built already
+        id: check_image
+        env:
+          DOCKERIMAGE: ${{ steps.image_hash.outputs.imagetag }}
+          DOCKERIMAGE_SLES: ${{ steps.image_hash.outputs.imagetag_sles }}
+        run: |
+          echo $DOCKER_TOKEN | docker login -u $DOCKER_USER --password-stdin
+
+          if [[ "$(docker manifest inspect $DOCKERIMAGE 2> /dev/null)" != "" ]]; then
+            echo "imageexists=true" >> $GITHUB_OUTPUT
+            echo "Image already exists, skip building available"
+          else
+            echo "imageexists=false" >> $GITHUB_OUTPUT
+            echo "Tag does not exist, build and publishing required"
+          fi
+
+          if [[ "$(docker manifest inspect $DOCKERIMAGE_SLES 2> /dev/null)" != "" ]]; then
+            echo "imageexists_sles=true" >> $GITHUB_OUTPUT
+            echo "SLES Image already exists, skip building available"
+          else
+            echo "imageexists_sles=false" >> $GITHUB_OUTPUT
+            echo "SLES Tag does not exist, build and publishing required"
+          fi
+
+
+  build_image:
+    name: Build image
+    runs-on: ROCM-Ubuntu
```

Comment:
check image stage runs on free runners and this on ROCM-Ubuntu. Both of them share docker images ? or is the image available for both ?

Reply:
The docker images gets pushed to the dockerhub, and then downloaded for jobs like Tidy, cppcheck, SLES.  In the case of the build_image job it needs to be on the ROCM-Ubuntu system in case it needs to build the image (takes 2 hours to build on free, 45 minutes on paid).  Once built the image has been built, it will be pushed to dockerhub and made available for other jobs

Reply:
in which step is it pulling the docker from docker hub ?